matrix/src/armory/matrix/matrix.py

```diff
@@ -0,0 +1,267 @@
+        keyword arguments of iterables, ranges, or sequences specifying the
+        allowable values for each argument. The yielded value of the generator
+        is a key-value mapping of arguments corresponding to a row of the
+        matrix.
+
+        If an argument to the generator is a callable, it will be called with
+        the preceding keyword parameter values for the current row. The return
+        of the callable is the value--or iterable--for the keyword, from which
+        to generate additional rows in the matrix.
+    """
+
+    if worker_num is None or num_workers is None:
+        worker_num = 0
+        num_workers = 1
+
+    def _generate(**kwargs) -> Iterable[Mapping[str, Any]]:
+        count = 0
+        for params in product({}, list(kwargs.items())):
+            # Skip over entries that are pruned
+            if prune is not None and prune(**params):
+                continue
+            # Skip over entries when partioning
+            if (count % num_workers) == worker_num:
+                yield params
+            count += 1
+
+    return _generate
+
+
+class Matrix:
+    """
+    A function to be invoked multiple times with parameters from rows of a
+    cartesian product matrix of all possible parameters.
+
+    This class is not intended to be used directly, but is used internally by
+    the `matrix` decorator.
+    """
+
+    def __init__(self, func: Callable, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.worker_num: Optional[int] = None
+        self.num_workers: Optional[int] = None
+        self.pruner: Optional[Callable[..., bool]] = None
+
+    @property
+    def matrix(self):
+        """The generated matrix of arguments"""
+        return create_matrix(
+            worker_num=self.worker_num,
+            num_workers=self.num_workers,
+            prune=self.pruner,
+        )(**self.kwargs)
+
+    def __call__(self, *args, **kwargs):
+        """
+        Invokes the function once for each row of the matrix. Any given keyword
+        arguments are merged with the keyword arguments from the matrix row.
+
+        Args:
+            *args: Positional arguments to be included with each function
+                invocation
+            **kwargs: Keyword arguments to be merged with the matrix row
+                arguments for each function invocation
+
+        Returns:
+            List of return values from each function invocation
+        """
+        results = []
+        for it in self.matrix:
+            it_args = deepcopy(args)
+            it_kwargs = deepcopy(kwargs)
+            it_kwargs.update(it)
+            try:
+                results.append(self.func(*it_args, **it_kwargs))
+            except Exception as err:
+                results.append(err)
+        return results
+
+    def override(self, **kwargs):
+        """Overrides the input parameter constraints for the matrix."""
+        self.kwargs.update(kwargs)
```

Comment:
So old overrides remain in force if the new `kwargs` doesn't contain previously specified parameters? Is this the right semantics?

Reply:
I considered two ways of handling functions like override, partition, etc.

1) (the current implementation) was that `override`, `partition`, & `filter` all operate in-place on the decorated function/`Matrix` object. This means you can write code like this
```python
@matrix(a=..., b=...)
def func(...):
    ...

func.override(a=...)
func() # uses original b, overridden a
func.override(b=...)
func() # uses overridden a and b
```

2) was that those functions returned a _copy_ of the `Matrix` object. This means you can write code like this
```python
func.override(a=...)
func() # still uses original a and b, you'd have to do `func2 = func.override(a=...)`

func.override(a=...)() # uses original b, overridden a
func.override(b=...)() # uses original a, overridden b
```

I don't have strong feelings one way or the other.

Reply:
The functional approach (option 2) seems preferable and has cleaner semantics. Matrix, including partition information, would be immutable, and the user applies operators to produce new Matrix objects.